File: asy/node/dirtree/dirtree.c
#include   <stdlib.h>
#include   <stdio.h>
#include   <dirent.h>
#include   <sys/stat.h>
#include   <errno.h>
#include   <string.h>

#include "tree.h"

/* 
 * output files or not: 
 *   0 means default, only output folders. 
 *   1 means output all 
 */
static int s_flag = 0;          /* 0 means default, only output folders. */

/* 
 * directory levels for output: -1 means no limit 
 */
static int s_max_level = -1;

const char head[] =
    "import fontsize;\n"
    "import \"../node/node.asy\" as node;\n\n"
    "settings.tex = \"xelatex\";\n"
    "texpreamble(\"\\usepackage{xeCJK}\");\n"
    "texpreamble(\"\\setCJKmainfont{SimHei}\");\n" 
    "texpreamble(\"\\setmonofont[Path=../fonts/]{andalemo.ttf}\");\n\n";

const char usage[] = "Usage: dirtree [-a] [-lxx] root_dir\n" 
                     "          -a    output both folders and files. Otherwise, only output folders by default.\n" 
                     "          -lxx  directory depth level. default no limit;\n";

TNODE* build_tree(char* dir, int flag);
int get_node_name(TNODE* node, char name[]);
inline int define_a_node(TNODE* node);
int attach_two_nodes(TNODE* dad, TNODE* kid);
void output_nodes(TNODE* root, int level);
char* get_last_part(char* path);
					 
#if (0)
#define DEBUG
#endif


int main(int argc, char *argv[])
{ 

	char rootname[100];
	char rootpath[1024];
	
	if(argc == 2){
		strcpy(rootpath, argv[1]);
	}
	else if(argc == 3){
		if(strncmp(argv[1], "-a", 2) == 0){
			s_flag = 1;
		}
		else if(strncmp(argv[1], "-l", 2) == 0){
			s_max_level = atoi(argv[1]+ 2);
		}
		else{
			printf("%s", usage);
			exit(0);
		}

		strcpy(rootpath, argv[2]);
	}
	else if(argc == 4){

		/* the order should be -a followed by -lxx */

		if(strcmp(argv[1], "-a") != 0){
			printf("%s", usage);
			exit(0);
		}
		s_flag = 1;
	
		if(strncmp(argv[2], "-l", 2) != 0){
			printf("%s", usage);
			exit(0);
		}
		s_max_level = atoi(argv[2] + 2);

		strcpy(rootpath, argv[3]);
	}
	else{
		printf("%s", usage);
		exit(0);
	}

#ifdef DEBUG
	printf("s_flag=%d\n", s_flag);
	printf("s_max_level=%d\n", s_max_level);
#endif

	
	/* delete the last '/' */
	if(rootpath[strlen(rootpath)-1] == '/'){
		rootpath[strlen(rootpath)-1] = '\0';
	}

	TNODE* root = build_tree(rootpath, s_flag);



        /********* draw the root tree *******/                             

	printf("%s\n\n", head);

	define_a_node(root);
	output_nodes(root, 1);

	get_node_name(root, rootname);
	printf("\n\n//change the following to draw_call_sequence() to produce call sequence.\n");
    printf("picture root = draw_dir_tree(%s);\n", rootname);
    //printf("attach(root.fit(), (0,0), SE);\n");
printf("attach(bbox(root, 2, 2, white), (0,0), SE);\n");

return 0;
}

/* recursive function: return the root node of the dir */
TNODE *
build_tree(char *root, int flag)
{
    DIR *dir;
    struct dirent *ptr;
    struct stat st;
    TNODE *troot;

#ifdef DEBUG
    printf("%s\n", root);
#endif

    /* todo: check if "root" is a dir */
    troot = tnode_new(NODE_TYPE_DIR);

    if (!troot)
        return NULL;

    /* get the last part of the full path */
    troot->txt = strdup(get_last_part(root));

    dir = opendir(root);
    if(!dir){
	printf("opendir(%s) failed: please make sure the directory name is correct!\n", root);
	exit(-1); 
    }

    while ((ptr = readdir(dir)) != NULL) {
        char full_path[2000];

        /* skip . and .. */
        if (strcmp(ptr->d_name, ".") == 0 || strcmp(ptr->d_name, "..") == 0) {
            continue;
        }

        strcpy(full_path, root);
        strcat(full_path, "/");
        strcat(full_path, ptr->d_name);

        stat(full_path, &st);

        if (S_ISDIR(st.st_mode)) {
            tnode_attach(troot, build_tree(full_path, flag));
        } else if (S_ISREG(st.st_mode)) {
#ifdef DEBUG
            printf("%s\n", ptr->d_name);
#endif
            if (flag) {
                TNODE *kid = tnode_new(NODE_TYPE_FILE);
                kid->txt = strdup(ptr->d_name);
                tnode_attach(troot, kid);
            } else {
                /* do nothing */
            }
        } else {
            printf("!!!!!!!!!!!!!not expected!!!!!!!!!!!!!!!!!!!\n");
        }
    }

    closedir(dir);

    return troot;
}

int
get_node_name(TNODE * node, char name[])
{
    int i, len;
    sprintf(name, "%s_%08x", node->txt, (unsigned int) node);

    /*
     * 1. replace '-' & '.' with '_' 
     * 2. if starting with digit, replace it with '_'
     */
    len = strlen(name);
    for (i = 0; i < len; i++) {
        if (name[i] == '-' || name[i] == '.')
            name[i] = '_';
    }

    if (isdigit(name[0]))
        name[0] = '_';

    return 0;
}

inline int
define_a_node(TNODE * node)
{
    char name[100];

    get_node_name(node, name);
    printf("node %-32s = node(\"%s\", \"%s\");\n", name, node->txt, node->type == NODE_TYPE_FILE ? "f" : "d");
}

int
attach_two_nodes(TNODE * dad, TNODE * kid)
{
    char dadname[100], kidname[100];
    get_node_name(dad, dadname);
    get_node_name(kid, kidname);
    printf("%s.attach(%s);\n", dadname, kidname);
    return 0;
}

void
output_nodes(TNODE * root, int level)
{
    TNODE *kid;

    if (s_max_level >= 0 && level > s_max_level)
        return;

    /* define all direct kids */
    if (NULL != (kid = root->kid)) {
        while (kid) {
            define_a_node(kid);
            kid = kid->sib;
        }
    }

    /* attach all direct kids */
    if (NULL != (kid = root->kid)) {
        while (kid) {
            attach_two_nodes(root, kid);
            kid = kid->sib;
        }
    }

    /* traverse on all direct kids */
    if (NULL != (kid = root->kid)) {
        while (kid) {
            output_nodes(kid, level + 1);
            kid = kid->sib;
        }
    }
}

char *
get_last_part(char *path)
{
    char *p;

    int len = strlen(path);

    for (p = path + len; p >= path; p--) {
        if (*p == '/')
            return p + 1;
    }

    return path;
}

File: asy/node/dirtree/dirtree.pl
#!/usr/bin/perl

use File::Basename;
use strict;


################################################################
# ref: "Perl 5 Complete", $7.5 example 3.
#
# notes on the data structure:

# here is a sample tree on file system:
#
# test
#  |
#  |- dir1
#  |   |
#  |   |- file1
#  |   \- file2
#  |- dir2
#  |   |
#  |   |- file3
#  |   \- file4
#  |- file5
#  \- file6
#
# it should be represented in perl as hash tables illustrated below:
#
#
#  {"test" => {"dir1" => {"file1"=>"FILE", 
#                         "file2"=>"FILE"
#                        },
#              "dir2" => {"file3"=>"FILE",
#	                  "file4"=>"FILE"
#			 },
#               "file5"=>"FILE",
#               "file6"=>"FILE"
#              }
#  }
#
################################################################

my $head = 
    "import fontsize;\n".
    "import \"../node/node.asy\" as node;\n\n". 
    "settings.tex = \"xelatex\";\n".
    "texpreamble(\"\\usepackage{xeCJK}\");\n".
    "texpreamble(\"\\setCJKmainfont{SimHei}\");\n".
    "texpreamble(\"\\setmonofont[Path=../fonts/]{andalemo.ttf}\");\n\n";

# hash value for file type keys are the same:
my $FILE = "__FILE__";  

################################################################
# it takes 3 arguments: path, level, and flag
# path: should be a directory, not accept a file
# level: 
#   - if level > 0, recursively call itself with level=(level-1);
#   - if level == 0, stop recursive calls.
# flag is either zero or non-zero:
#   - 0 means dir only
#   - otherwise, means dir and files
#
# it returns a ref to a hash contains the entries under path, the
# basename of the path itself is not included.
################################################################
sub build_tree
{
	my ($path, $level, $flag) = @_;
	my ($file, $kids) = ('', {});
	
	# sanity checks
	if(!(-e $path)){
		die "$path does not exist.\n";
	}

	if(!(-d $path)){
		die "$path is not a directory\n";
	}
		
	opendir(FD, $path);
	my @files = readdir(FD);
	closedir(FD);
		
	chdir($path);

	foreach $file (@files){

		next if ($file eq '.' || $file eq '..');

		if(-f $file){
			if($flag != 0){
				$kids->{$file} = $FILE;
			}
		}
		elsif (-d $file){

			if($level > 0){
				$level -= 1;
				$kids->{$file} = build_tree($file, $level, $flag);
			}
		}
	}

	chdir("..");
	
	#$dad->{basename($path)} = $kids;
	#return $dad;
	return $kids;
}

###############################################################
# it takes one argument: the REF to the dir or file name
# it converts a dir/file name into an unique node name which also
# obeys asymptote variable naming rules
###############################################################
sub get_node_asy_name
{
	my ($name) = (@_);
	my $node_name = $$name.'_'.$name;
	$node_name =~ s/SCALAR|HASH//g;
	$node_name =~ s/[-+.&() ]/_/g;
	return $node_name;
}

###############################################################
# it takes 2 arguments: the key and the value of the ndoe
# it outputs the node definition in asymptote, such as:
#   node my_node = node("my node name", "f");
###############################################################
sub define_asy_node
{
	my ($key, $value) = (@_);
	my $type;

	if($value eq $FILE){
		$type = "f";
	}
	else{
		$type = "d";
	}

	print "node ";
        print get_node_asy_name(\$key);
	print " = node(\"".$key."\", \"";
	print "$type\");\n";
}
		

###############################################################
# it takes 2 arguments: the REFs to the dad and kid node name
###############################################################
sub attach_asy_node
{
	my ($dad, $kid) = (@_);
	print get_node_asy_name($dad);
	print ".attach(";
	print get_node_asy_name($kid);
	print ");\n";
}

################################################################
# it takes 2 argument: REF to the root key, and
# REF to the value of the root key:
#  - $rootkey: its assumed that the asy node of the rootkey is already created,
#    it's provided here for its kids to attach to it.
#  - $rootvalue: hashref to the content of the root. cann't be $FILE, as the
#    sanity test excludes this case
#
################################################################
sub output_tree
{
	my ($dad, $kids) = @_;
	my $dad_node;
	my $key;

	$dad_node = get_node_asy_name(\$dad);

	foreach $key (keys %$kids){
		define_asy_node(\$key);

		if(ref $kids->{$key}){
			output_tree(\$key, $kids->{$key});
		}
		else{
			attach_asy_node(\$dad, \$key);
		}
	}
}
my $tree = {};
$tree->{"test"} = build_tree("test", 10, 1);
output_tree($tree);

